File: Public/Gui/Linn/Gui/Resources/Controller.py
import sys
import wx
import Linn.Gui.Renderer
import Linn.Gui.Resources.Plugin
import Linn.Gui.Resources.Message
import Linn.Gui.Serialize
import Linn.Gui.SceneGraph.Srt

class Timer(wx.Timer):
    def __init__(self, aParent=None):
        wx.Timer.__init__(self)
        self.iParent = aParent
        
    def __deepcopy__(self, memo={}):
        new = self.__class__()
        new.iParent = memo[id(self.iParent)]
        return new
        
    def Notify(self):
        if self.iParent:
            self.iParent.Notify()
            Linn.Gui.Renderer.GetRenderer().Render()
    
        
class Monostable(Linn.Gui.Resources.Plugin.Plugin):
    def __init__(self, aName=None, aPeriod=0.0):
        if aName == None:
            aName = self.UniqueName('Monostable')
        Linn.Gui.Resources.Plugin.Plugin.__init__(self, aName)
        self.iTimer = Timer(self)
        self.iState = False
        self.iPeriod = aPeriod
    
    def Notify(self):
        self.SetState(False)
        
    def Period(self):
        return self.iPeriod
    
    def SetPeriod(self, aPeriod):
        self.iPeriod = aPeriod
        self.ObserverUpdate()
        
    def SetState(self, aState):
        if aState != self.iState:
            self.iState = aState
            if self.iPlugin:
                self.iPlugin.Vector3d(Linn.Gui.SceneGraph.Srt.Vector3d(int(self.iState), 0, 0))
            msg = Linn.Gui.Resources.Message.MsgStateChanged(self.Fullname(), not self.iState, self.iState)
            self.SendMessage(msg)
        
    def Hit(self, aVector3d):
        self.SetState(True)

    def UnHit(self):
        if self.iPeriod:
            self.iTimer.Start(self.iPeriod*1000, wx.TIMER_ONE_SHOT)
        else:
            self.iTimer.Start(1, wx.TIMER_ONE_SHOT)
                
    def Load(self, aStream):
        Linn.Gui.Resources.Plugin.Plugin.Load(self, aStream)
        
        period = aStream.LoadNode('Period')
        if len(period) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<Period>', len(period), 1)
        self.iPeriod = float(period[0].firstChild.nodeValue)
        #print 'period =', self.iPeriod
    
    def Save(self, aStream):
        if aStream.ContextAttribute('class') == None:
            aStream.SetContextAttribute('class', 'Monostable')
            
        Linn.Gui.Resources.Plugin.Plugin.Save(self, aStream)
        
        period = aStream.CreateElement('Period')
        period.appendChild(aStream.CreateTextNode(str(self.iPeriod)))
        aStream.SaveNode(period)
                    

class Bistable(Linn.Gui.Resources.Plugin.Plugin):
    def __init__(self, aName=None, aStartState=0):
        if aName == None:
            aName = self.UniqueName('Bistable')
        Linn.Gui.Resources.Plugin.Plugin.__init__(self, aName)
        self.iStartState = False
        self.iState = False
        
    def StartState(self):
        return self.iStartState
    
    def SetStartState(self, aStartState):
        self.iStartState = aStartState
        self.ObserverUpdate()
        
    def SetState(self, aState):
        if aState != self.iState:
            self.iState = aState
            if self.iPlugin:
                self.iPlugin.Vector3d(Linn.Gui.SceneGraph.Srt.Vector3d(int(self.iState), 0, 0))
            msg = Linn.Gui.Resources.Message.MsgStateChanged(self.Fullname(), not self.iState, self.iState)
            self.SendMessage(msg)
        
    def Hit(self, aVector3d):
        self.SetState(not self.iState)
    
    def ReceiveMessage(self, aMessage):
        if aMessage.Fullname() == self.Fullname():
            if isinstance(aMessage, Linn.Gui.Resources.Message.MsgSetState):
                self.SetState(aMessage.Value())
            if isinstance(aMessage, Linn.Gui.Resources.Message.MsgToggleState):
                self.SetState(not self.iState)
            return
        Linn.Gui.Resources.Plugin.Plugin.ReceiveMessage(self, aMessage)
    
    def Load(self, aStream):
        Linn.Gui.Resources.Plugin.Plugin.Load(self, aStream)
        
        state = aStream.LoadNode('StartState')
        if len(state) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<StartState>', len(state), 1)
        value = state[0].firstChild.nodeValue.lower()
        if(value == '1' or value == 'true'):
            self.iStartState = True
        else:
            self.iStartState = False
        self.iState = self.iStartState
        #print 'startState =', self.iStartState
    
    def Save(self, aStream):
        if aStream.ContextAttribute('class') == None:
            aStream.SetContextAttribute('class', 'Bistable')
            
        Linn.Gui.Resources.Plugin.Plugin.Save(self, aStream)
        
        state = aStream.CreateElement('StartState')
        state.appendChild(aStream.CreateTextNode(str(self.iStartState).lower()))
        aStream.SaveNode(state)
        
        
class Counter(Linn.Gui.Resources.Plugin.Plugin):
    def __init__(self, aName=None, aCountMax=0, aCountsPerSecond=0, aLoop=False):
        if aName == None:
            aName = self.UniqueName('Counter')
        Linn.Gui.Resources.Plugin.Plugin.__init__(self, aName)
        self.iTimer = Timer(self)
        self.iCount = 0
        self.iCountMax = aCountMax
        if aCountsPerSecond:
            self.iTimeStep = 1.0 / aCountsPerSecond
        else:
            self.iTimeStep = 0.0
        self.iCounting = False
        self.iReverseCount = 0
        self.iLoop = aLoop
        
    def MaxCount(self):
        return self.iCountMax
    
    def SetMaxCount(self, aMaxCount):
        self.iCountMax = aMaxCount
        if self.iCount >= self.iCountMax:
            self.SetCount(self.iCountMax - 1)
        self.ObserverUpdate()
        
    def Count(self):
        return self.iCount
    
    def SetCount(self, aCount):
        if aCount < 0:
            self.iCount = 0
        elif aCount >= self.iCountMax:
            self.iCount = self.iCountMax - 1
        else:
            self.iCount = aCount
        if self.iPlugin:
            self.iPlugin.Vector3d(Linn.Gui.SceneGraph.Srt.Vector3d(self.iCount, 0.0, 0.0))
        self.ObserverUpdate()
        
    def Fps(self):
        if self.iTimeStep:
            return 1.0 / self.iTimeStep
        else:
            return self.iTimeStep
        
    def SetFps(self, aFps):
        if aFps > 0:
            self.iTimeStep = 1.0 / aFps
        else:
            self.iTimeStep = 0
        self.ObserverUpdate()
        
    def Loop(self):
        return self.iLoop
    
    def SetLoop(self, aLoop):
        self.iLoop = aLoop
        if self.iLoop:
            self.Start()
        self.ObserverUpdate()
        
    def Counting(self):
        return self.iCounting
        
    def Start(self):
        self.iCounting = True
        if self.iTimeStep:
            self.iTimer.Start(self.iTimeStep*1000, wx.TIMER_ONE_SHOT)
        self.ObserverUpdate()
        
    def Stop(self):
        self.iCounting = False
        self.iTimer.Stop()
        self.ObserverUpdate()
            
    def Notify(self):
        counts = 1    # we have no count length so just update as fast as possible
        # subtract frames from index if counting in reverse
        count = self.iCount - (~counts + ((2 * self.iReverseCount * counts) + 1))
        newCount = count % self.iCountMax
        if newCount == count or self.iLoop:
            # inc/dec or wrap the counter
            self.iCount = newCount
            if self.iTimeStep:
                self.iTimer.Start(self.iTimeStep*1000, wx.TIMER_ONE_SHOT)
        else:
            # clamp counter to bottom/top depending on count direction
            self.iCount = (not self.iReverseCount) * (self.iCountMax - 1)
            self.iCounting = False
            msg = Linn.Gui.Resources.Message.MsgCountEnd(self.Fullname())
            self.SendMessage(msg)
        if self.iPlugin:
            self.iPlugin.Vector3d(Linn.Gui.SceneGraph.Srt.Vector3d(self.iCount, 0.0, 0.0))
        self.ObserverUpdate()
        
    def Vector3d(self, aVector3d):
        self.iReverseCount = not aVector3d.iX
        self.Start()
            
    def ReceiveMessage(self, aMessage):
        if aMessage.Fullname() == self.Fullname():
            if isinstance(aMessage, Linn.Gui.Resources.Message.MsgCountStart):
                self.Start()
            if isinstance(aMessage, Linn.Gui.Resources.Message.MsgCountStop):
                self.Stop()
            return
        Linn.Gui.Resources.Plugin.Plugin.ReceiveMessage(self, aMessage)
            
    def Load(self, aStream):
        Linn.Gui.Resources.Plugin.Plugin.Load(self, aStream)
        
        count = aStream.LoadNode('MaxCount')
        if len(count) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<MaxCount>', len(count), 1)
        self.iCountMax = int(count[0].firstChild.nodeValue)
        #print 'countMax =', self.iCountMax
        
        cps = aStream.LoadNode('CountsPerSecond')
        if len(cps) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<CountsPerSecond>', len(cps), 1)
        countsPerSecond = float(cps[0].firstChild.nodeValue)
        self.SetFps(countsPerSecond)
        #print 'cps =', self.iTimeStep
        
        loop = aStream.LoadNode('Loop')
        if len(cps) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<Loop>', len(loop), 1)
        value = loop[0].firstChild.nodeValue.lower()
        if(value == '1' or value == 'true'):
            self.iLoop = True
        else:
            self.iLoop = False
        if self.iLoop:
            self.iCounting = True
            if self.iTimeStep:
                self.iTimer.Start(self.iTimeStep*1000, wx.TIMER_ONE_SHOT)
        #print 'loop =', self.iLoop
    
    def Save(self, aStream):
        if aStream.ContextAttribute('class') == None:
            aStream.SetContextAttribute('class', 'Counter')
            
        Linn.Gui.Resources.Plugin.Plugin.Save(self, aStream)
        
        count = aStream.CreateElement('MaxCount')
        count.appendChild(aStream.CreateTextNode(str(int(self.iCountMax))))
        aStream.SaveNode(count)
        
        cps = aStream.CreateElement('CountsPerSecond')
        if self.iTimeStep:
            cps.appendChild(aStream.CreateTextNode(str(1.0 / self.iTimeStep)))
        else:
            cps.appendChild(aStream.CreateTextNode(str(0)))
        aStream.SaveNode(cps)
        
        loop = aStream.CreateElement('Loop')
        loop.appendChild(aStream.CreateTextNode(str(self.iLoop).lower()))
        aStream.SaveNode(loop)
    
    
    
            

File: Gui/Linn/Gui/Resources/Controller.cs
using Linn.Gui.Resources;
using System.Xml;
using Linn.Gui.Scenegraph;
using System;
using Linn;
using System.Globalization;

namespace Linn {
namespace Gui {
namespace Resources {
    
public sealed class Monostable : Plugin
{
    public Monostable() : base("Monostable") {
        iTimer = new Timer();
        iTimer.Elapsed += Notify;
        iTimer.AutoReset = false;
    }
    
    public Monostable(string aName) : base(aName) {
        iTimer = new Timer();
        iTimer.Elapsed += Notify;
        iTimer.AutoReset = false;
    }
    
    public override void Load(XmlNode aXmlNode) {
        base.Load(aXmlNode);
        
        XmlNodeList list;
        list = aXmlNode.SelectNodes("Period");
        if(list != null) {
            Assert.Check(list.Count == 1);
            NumberFormatInfo nfi = new NumberFormatInfo();
            nfi.NumberDecimalSeparator = ".";
            iPeriod = float.Parse(list[0].FirstChild.Value, nfi);
        }
    }
    
    public override void Save(XmlTextWriter aWriter) {
        base.Save(aWriter);
        
        aWriter.WriteStartElement("Period");
        NumberFormatInfo nfi = new NumberFormatInfo();
        nfi.NumberDecimalSeparator = ".";
        aWriter.WriteString(iPeriod.ToString(nfi));
        aWriter.WriteEndElement();
    }
    
    public float Period {
        get {
            return iPeriod;
        }
        set {
            iPeriod = value;
            if(iPeriod > 0) {
                iTimer.Interval = iPeriod * 1000;
            }
            ObserverUpdate();
        }
    }
    
    private void Notify(object aSender, EventArgs aArgs) {
        iTimer.Stop();
        State = false;
        Renderer.Instance.Render();
    }
    
    public override void Hit(Vector3d aVector) {
        iTimer.Stop();
        State = true;
    }
    
    public override void UnHit() {
        if(iPeriod > 0.0f) {
            iTimer.Start();
        } else {
            State = false;
        }
    }
    
    public override bool ProcessMessage(Message aMessage) {
        if(aMessage.Fullname == Fullname) {
            MsgSetState setState = aMessage as MsgSetState;
            if(setState != null) { 
                State = setState.State;
            }
            return true;
        }
        if(base.ProcessMessage(aMessage)) {
            return true;
        }
        return false;
    }
    
    public bool State {
        get {
            return iState;
        }
        set{
            if(iState != value) {
                iState = value;
                if(NextPlugin != null) {
                    if(iState == true) {
                        NextPlugin.Vector3d(new Vector3d(1, 0, 0));
                    } else {
                        NextPlugin.Vector3d(new Vector3d(0, 0, 0));
                    }
                }
                MsgStateChanged msg = new MsgStateChanged(this, !iState, iState);
                SendMessage(msg);
            }
        }
    }
    
    private bool iState = false;
    private float iPeriod = 0.0f;
    private Timer iTimer = null;
}

public sealed class Bistable : Plugin
{
    public Bistable() : base("Bistable") {
    }
    
    public Bistable(string aName) : base(aName) {
    }
    
    public override void Load(XmlNode aXmlNode) {
        base.Load(aXmlNode);
        
        XmlNodeList list;
        list = aXmlNode.SelectNodes("StartState");
        if(list != null) {
            Assert.Check(list.Count == 1);
            iState = bool.Parse(list[0].FirstChild.Value);
        }
    }
    
    public override void Link() {
        base.Link();
        if(NextPlugin != null) {
            if(iState == true) {
                NextPlugin.Vector3d(new Vector3d(1, 0, 0));
            } else {
                NextPlugin.Vector3d(new Vector3d(0, 0, 0));
            }
        }
    }
    
    public override void Save(XmlTextWriter aWriter) {
        base.Save(aWriter);
        
        aWriter.WriteStartElement("StartState");
        aWriter.WriteString(iState.ToString());
        aWriter.WriteEndElement();
    }
    
    public override void Hit(Vector3d aVector) {
        State = !iState;
    }
    
    public override bool ProcessMessage(Message aMessage) {
        if(aMessage.Fullname == Fullname) {
            MsgSetState setState = aMessage as MsgSetState;
            if(setState != null) {
                State = setState.State;
                return true;
            }
            MsgToggleState toggleState = aMessage as MsgToggleState;
            if(toggleState != null) {
                State = !State;
                return true;
            }
        }
        if(base.ProcessMessage(aMessage)) {
            return true;
        }
        return false;
    }
    
    public bool State {
        get {
            return iState;
        }
        set {
            if(iState != value) {
                iState = value;
                if(NextPlugin != null) {
                    if(iState == true) {
                        NextPlugin.Vector3d(new Vector3d(1, 0, 0));
                    } else {
                        NextPlugin.Vector3d(new Vector3d(0, 0, 0));
                    }
                }
                MsgStateChanged msg = new MsgStateChanged(this, !iState, iState);
                SendMessage(msg);
                ObserverUpdate();
            }
        }
    }
    
    private bool iState = false;
}

public sealed class Counter : Plugin
{
    public Counter() : base("Counter") {
        iTimer = new Timer();
        iTimer.Interval = 40;
        iTimer.Elapsed += Notify;
        iTimer.AutoReset = false;
    }
    
    public Counter(string aName) : base(aName) {
        iTimer = new Timer();
        iTimer.Interval = 40;
        iTimer.Elapsed += Notify;
        iTimer.AutoReset = false;
    }
    
    public override void Load(XmlNode aXmlNode) {
        base.Load(aXmlNode);
        
        XmlNodeList list;
        list = aXmlNode.SelectNodes("MaxCount");
        if(list != null) {
            Assert.Check(list.Count == 1);
            iMaxCount = int.Parse(list[0].FirstChild.Value);
        }
        
        list = aXmlNode.SelectNodes("Count");
        if(list != null) {
            if(list.Count > 0) {
                Assert.Check(list.Count == 1);
                iCount = int.Parse(list[0].InnerText);
            }
        }
        
        list = aXmlNode.SelectNodes("CountsPerSecond");
        if(list != null) {
            Assert.Check(list.Count == 1);
            NumberFormatInfo nfi = new NumberFormatInfo();
            nfi.NumberDecimalSeparator = ".";
            CountsPerSecond = float.Parse(list[0].FirstChild.Value, nfi);
        }
        
        list = aXmlNode.SelectNodes("Loop");
        if(list != null) {
            Assert.Check(list.Count == 1);
            Loop = bool.Parse(list[0].FirstChild.Value);
        }

        list = aXmlNode.SelectNodes("Counting");
        if(list != null) {
            Assert.Check(list.Count == 1);
            iCounting = bool.Parse(list[0].FirstChild.Value);
            if(iCounting == true) {
                Start();
            } else {
                Stop();
            }
        }
    }
    
    public override void Link() {
        base.Link();
        if(NextPlugin != null) {
            NextPlugin.Vector3d(new Vector3d(iCount, 0, 0));
        }
    }
    
    public override void Save(XmlTextWriter aWriter) {
        base.Save(aWriter);
        
        aWriter.WriteStartElement("MaxCount");
        aWriter.WriteString(iMaxCount.ToString());
        aWriter.WriteEndElement();
        
        aWriter.WriteStartElement("CountsPerSecond");
        aWriter.WriteString(iCountsPerSecond.ToString());
        aWriter.WriteEndElement();
        
        aWriter.WriteStartElement("Loop");
        aWriter.WriteString(iLoop.ToString());
        aWriter.WriteEndElement();
        
        aWriter.WriteStartElement("Counting");
        aWriter.WriteString(iCounting.ToString());
        aWriter.WriteEndElement();
    }
    
    public void Notify(object aSender, EventArgs aArgs) {
        long curTime = DateTime.Now.Ticks;
        float elapsedTime = (float)TimeSpan.FromTicks(curTime - iLastTime).TotalSeconds;
        int count = (int)Math.Floor((elapsedTime + iRemainder) / iTimeStep);
        iRemainder = (elapsedTime + iRemainder) - (iTimeStep * count);
        System.Console.WriteLine("elapsedTime=" + elapsedTime + ", iTimeStep=" + iTimeStep + ", count=" + count + ", iRemainder=" + iRemainder);
        // subtract frames from index if counting in reverse
        if(iReverseCount) {
            count = Count - count;
        } else {
            count = Count + count;
        }
        int newCount = count % iMaxCount;
        Trace.WriteLine(Trace.kGui, "newCount = " + newCount + ", count = " + count);
        if((newCount == count && count >= 0) || Loop) {
            // inc/dec or wrap the counter
            Count = newCount;
            iLastTime = curTime;
            iTimer.Start();
        } else {
            // clamp counter to bottom/top depending on count direction
            if(iReverseCount) {
                Count = 0;
            } else {
                Count = iMaxCount - 1;
            }
            Stop();
            MsgCountEnd msg = new MsgCountEnd(this);
            SendMessage(msg);
        }
        Renderer.Instance.Render();
    }
    
    public float CountsPerSecond {
        get {
            return iCountsPerSecond;
        }
        set {
            if(value != iCountsPerSecond) {
                iCountsPerSecond = value;
                if(iCountsPerSecond > 0) {
                    iTimeStep = 1.0f / iCountsPerSecond;
                } else {
                    Stop();
                }
                ObserverUpdate();
            }
        }
    }
    
    public bool Loop {
        get {
            return iLoop;
        }
        set {
            if(value != iLoop) {
                iLoop = value;
                /*if(iLoop) {
                    if(iTimeStep > 0 && !iCounting) {
                        Start();
                    }
                } else {
                    Stop();
                }*/
                ObserverUpdate();
            }
        }
    }
    
    public int MaxCount {
        get {
            return iMaxCount;
        }
        set {
            if(value != iMaxCount) {
                iMaxCount = value;
                if(iCount >= iMaxCount) {
                    Count = iMaxCount - 1;
                }
                ObserverUpdate();
            }
        }
    }
    
    public int Count {
        get {
            return iCount;
        }
        set {
            if(value != iCount) {
                Trace.WriteLine(Trace.kGui, "iCount = " + iCount);
                if(value < 0) {
                    iCount = 0;
                } else if(value >= iMaxCount) {
                    iCount = iMaxCount - 1;
                } else {
                    iCount = value;
                }
                if(NextPlugin != null) {
                    NextPlugin.Vector3d(new Vector3d(iCount, 0, 0));
                }
                ObserverUpdate();
            }
        }
    }
    
    public bool Counting {
        get {
            return iCounting;
        }
    }
    
    public void Start() {
        Trace.WriteLine(Trace.kGui, "Counter's timer started.");
        iCounting = true;
        iTimer.Start();
        iLastTime = DateTime.Now.Ticks;
    }
    
    public void Stop() {
        Trace.WriteLine(Trace.kGui, "Counter's timer stopped.");
        iCounting = false;
        iReverseCount = false;
        iTimer.Stop();
    }
    
    public override void Vector3d(Vector3d aVector) {
        if(aVector.X == 0) {
            iReverseCount = true;
        } else {
            iReverseCount = false;
        }
        if(iTimeStep > 0 && !iCounting) {
            Start();
        }
    }
    
    public override bool ProcessMessage(Message aMessage) {
        if(aMessage.Fullname == Fullname) {
            MsgCountStart start = aMessage as MsgCountStart;
            if(start != null) {
                Start();
                return true;
            }
            MsgCountStop stop = aMessage as MsgCountStop;
            if(stop != null) {
                Stop();
                return true;
            }
        }
        if(base.ProcessMessage(aMessage)) {
            return true;
        }
        return false;
    }
        
    private float iCountsPerSecond = 0.0f;
    private bool iLoop = false;
    private int iMaxCount = 0;
    private int iCount = 0;
    private float iTimeStep = 0;
    private bool iCounting = false;
    private bool iReverseCount = false;
    private Timer iTimer = null;
    private long iLastTime;
    private float iRemainder = 0.0f;
}

} // Resources
} // Gui
} // Linn
